crates/bitwarden-crypto/src/keys/symmetric_crypto_key.rs

```diff
@@ -10,46 +11,95 @@ use rand::SeedableRng;
 pub struct Aes256CbcKey {
     /// Uses a pinned heap data structure, as noted in [Pinned heap data][crate#pinned-heap-data]
     pub(crate) enc_key: Pin<Box<GenericArray<u8, U32>>>,
 }
 
-/// Aes256CbcHmacKey is a symmetric encryption key consisting
+impl ConstantTimeEq for Aes256CbcKey {
+    fn ct_eq(&self, other: &Self) -> Choice {
+        self.enc_key.ct_eq(&other.enc_key)
+    }
+}
+
+impl PartialEq for Aes256CbcKey {
+    fn eq(&self, other: &Self) -> bool {
+        self.ct_eq(other).into()
+    }
+}
+
+/// [Aes256CbcHmacKey] is a symmetric encryption key consisting
 /// of two 256-bit keys, one for encryption and one for MAC
 #[derive(ZeroizeOnDrop, Clone)]
-#[cfg_attr(test, derive(Debug, PartialEq))]
 pub struct Aes256CbcHmacKey {
     /// Uses a pinned heap data structure, as noted in [Pinned heap data][crate#pinned-heap-data]
     pub(crate) enc_key: Pin<Box<GenericArray<u8, U32>>>,
     /// Uses a pinned heap data structure, as noted in [Pinned heap data][crate#pinned-heap-data]
     pub(crate) mac_key: Pin<Box<GenericArray<u8, U32>>>,
 }
 
+impl ConstantTimeEq for Aes256CbcHmacKey {
+    fn ct_eq(&self, other: &Self) -> Choice {
+        self.enc_key.ct_eq(&other.enc_key) & self.mac_key.ct_eq(&other.mac_key)
+    }
+}
+
+impl PartialEq for Aes256CbcHmacKey {
+    fn eq(&self, other: &Self) -> bool {
+        self.ct_eq(other).into()
+    }
+}
+
+#[derive(Zeroize, Clone)]
+#[cfg_attr(test, derive(Debug))]
+pub struct XChaCha20Poly1305Key {
+    pub(crate) key_id: [u8; 24],
+    pub(crate) enc_key: Pin<Box<GenericArray<u8, U32>>>,
+}
+
+impl ConstantTimeEq for XChaCha20Poly1305Key {
+    fn ct_eq(&self, other: &Self) -> Choice {
+        self.enc_key.ct_eq(&other.enc_key)
+    }
+}
+
+impl PartialEq for XChaCha20Poly1305Key {
+    fn eq(&self, other: &Self) -> bool {
+        self.ct_eq(other).into()
+    }
+}
+
 /// A symmetric encryption key. Used to encrypt and decrypt [`EncString`](crate::EncString)
 #[derive(ZeroizeOnDrop, Clone)]
-#[cfg_attr(test, derive(PartialEq))]
 pub enum SymmetricCryptoKey {
     Aes256CbcKey(Aes256CbcKey),
     Aes256CbcHmacKey(Aes256CbcHmacKey),
+    // always encode with cose
+    XChaCha20Poly1305Key(XChaCha20Poly1305Key),
 }
 
 impl SymmetricCryptoKey {
-    const KEY_LEN: usize = 32;
-    const MAC_LEN: usize = 32;
-
-    /// Generate a new random [SymmetricCryptoKey]
-    pub fn generate(mut rng: impl rand::RngCore) -> Self {
+    // enc type 0 old static format
+    const AES256_CBC_KEY_LEN: usize = 32;
+    // enc type 2 old static format
+    const AES256_CBC_HMAC_KEY_LEN: usize = 64;
+
+    /// Generate a new random AES256_CBC [SymmetricCryptoKey]
+    ///
+    /// WARNING: This function should only be used with a proper cryptographic RNG. If you do not
+    /// have a good reason for using this function, use
+    /// [SymmetricCryptoKey::generate_aes256_cbc_hmac] instead.
+    pub(crate) fn generate_aes256_cbc_hmac_internal(rng: &mut impl Rng) -> Self {
```

Comment:
Is the change to `&mut impl Rng` needed? I think the older `mut rng: impl rand::RngCore` would also work. We should try to be consistent with how we use it everywhere.

Reply:
Agreed, reverted but added `+ rand::CryptoRng` to it.